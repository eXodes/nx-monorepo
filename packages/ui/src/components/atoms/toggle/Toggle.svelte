<script lang="ts">
import { classNames } from "@nx-monorepo/utils";
import { createEventDispatcher } from "svelte";
import { Switch } from "@rgossiaux/svelte-headlessui";

export let label: string;
export let enabled = false;

const dispatch = createEventDispatcher<{
  change: boolean;
}>();

const handleChange = (bool: boolean) => {
  enabled = bool;
  dispatch("change", bool);
};
</script>

<Switch
  checked={enabled}
  on:change={(e) => handleChange(e.detail)}
  class="bg-brand-400 focus:ring-brand-500 relative inline-flex h-8 w-16 flex-shrink-0 cursor-pointer rounded-full border-2 border-transparent transition-colors duration-200 ease-in-out focus:outline-none focus:ring-2 focus:ring-offset-2 dark:ring-offset-gray-900"
  data-testid="toggle"
>
  <span class="sr-only">{label}</span>
  <span
    class={classNames(
      enabled ? "translate-x-7" : "translate-x-0",
      "pointer-events-none relative inline-block h-7 w-8 rounded-full bg-white dark:bg-gray-900 shadow transform ring-0 transition ease-in-out duration-200"
    )}
  >
    <span
      class={classNames(
        enabled
          ? "opacity-0 ease-out duration-100"
          : "opacity-100 ease-in duration-200",
        "absolute inset-0 h-full w-full flex items-center justify-center transition-opacity"
      )}
      aria-hidden="true"
    >
      <slot name="disabled" />
    </span>
    <span
      class={classNames(
        enabled
          ? "opacity-100 ease-in duration-200"
          : "opacity-0 ease-out duration-100",
        "absolute inset-0 h-full w-full flex items-center justify-center transition-opacity"
      )}
      aria-hidden="true"
    >
      <slot name="enabled" />
    </span>
  </span>
</Switch>
