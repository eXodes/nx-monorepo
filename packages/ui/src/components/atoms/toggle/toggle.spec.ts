import Toggle from "./Toggle.svelte";
import { render, fireEvent } from "@testing-library/svelte";

describe("Toggle", () => {
  it("renders", () => {
    const { container } = render(Toggle, {
      props: {
        label: "Toggle",
      },
    });

    expect(container).toMatchSnapshot();
  });

  it("renders with checked props 'true'", async () => {
    const { container, getByTestId } = render(Toggle, {
      props: {
        enabled: true,
        label: "Toggle",
      },
    });

    const element = getByTestId("toggle");

    expect(container).toMatchSnapshot();
    expect(element.getAttribute("aria-checked")).toBe("true");
  });

  it("it should toggle checked value to 'true' when clicked once", async () => {
    const { getByTestId } = render(Toggle, {
      props: {
        label: "Toggle",
      },
    });

    const element = getByTestId("toggle");

    await fireEvent.click(element);

    expect(element.getAttribute("aria-checked")).toBe("true");
  });

  it("it should toggle checked value to 'false' when clicked twice", async () => {
    const { getByTestId } = render(Toggle, {
      props: {
        label: "Toggle",
      },
    });

    const element = getByTestId("toggle");

    await fireEvent.click(element);
    await fireEvent.click(element);

    expect(element.getAttribute("aria-checked")).toBe("false");
  });
});
