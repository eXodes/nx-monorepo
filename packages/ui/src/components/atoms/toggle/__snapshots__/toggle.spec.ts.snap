// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle renders 1`] = `
<body>
  <div>
    <button
      aria-checked="false"
      class="bg-brand-400 focus:ring-brand-500 relative inline-flex h-8 w-16 flex-shrink-0 cursor-pointer rounded-full border-2 border-transparent transition-colors duration-200 ease-in-out focus:outline-none focus:ring-2 focus:ring-offset-2 dark:ring-offset-gray-900"
      data-testid="toggle"
      id="headlessui-switch-1"
      role="switch"
      tabindex="0"
      type="button"
    >
      <span
        class="sr-only"
      >
        Toggle
      </span>
       
      <span
        class="translate-x-0 pointer-events-none relative inline-block h-7 w-8 rounded-full bg-white dark:bg-gray-900 shadow transform ring-0 transition ease-in-out duration-200"
      >
        <span
          aria-hidden="true"
          class="opacity-100 ease-in duration-200 absolute inset-0 h-full w-full flex items-center justify-center transition-opacity"
        />
         
        <span
          aria-hidden="true"
          class="opacity-0 ease-out duration-100 absolute inset-0 h-full w-full flex items-center justify-center transition-opacity"
        />
      </span>
    </button>
    
    
    
  </div>
</body>
`;

exports[`Toggle renders with checked props 'true' 1`] = `
<body>
  <div>
    <button
      aria-checked="true"
      class="bg-brand-400 focus:ring-brand-500 relative inline-flex h-8 w-16 flex-shrink-0 cursor-pointer rounded-full border-2 border-transparent transition-colors duration-200 ease-in-out focus:outline-none focus:ring-2 focus:ring-offset-2 dark:ring-offset-gray-900"
      data-testid="toggle"
      id="headlessui-switch-2"
      role="switch"
      tabindex="0"
      type="button"
    >
      <span
        class="sr-only"
      >
        Toggle
      </span>
       
      <span
        class="translate-x-7 pointer-events-none relative inline-block h-7 w-8 rounded-full bg-white dark:bg-gray-900 shadow transform ring-0 transition ease-in-out duration-200"
      >
        <span
          aria-hidden="true"
          class="opacity-0 ease-out duration-100 absolute inset-0 h-full w-full flex items-center justify-center transition-opacity"
        />
         
        <span
          aria-hidden="true"
          class="opacity-100 ease-in duration-200 absolute inset-0 h-full w-full flex items-center justify-center transition-opacity"
        />
      </span>
    </button>
    
    
    
  </div>
</body>
`;
